src/repositories/updater.rs

```diff
@@ -0,0 +1,354 @@
+use crate::error::Result;
+use crate::repositories::{GitHub, GitLab};
+use crate::utils::{daemon::cron, MetadataPackage};
+use crate::{db::Pool, Config, Context};
+use chrono::{DateTime, Utc};
+use log::{debug, info, trace, warn};
+use once_cell::sync::Lazy;
+use postgres::Client;
+use regex::Regex;
+use std::collections::{HashMap, HashSet};
+use std::fmt;
+use std::sync::Arc;
+use std::time::Duration;
+
+pub trait RepositoryForge {
+    /// Result used both as the `host` column in the DB and to match repository URLs during
+    /// backfill.
+    fn host(&self) -> &str;
+
+    /// FontAwesome icon used in the front-end.
+    fn icon(&self) -> &'static str;
+
+    /// How many items we can query in one graphql request.
+    fn chunk_size(&self) -> usize;
+
+    /// Used by both backfill_repositories and load_repository. When the repository is missing
+    /// `None` is returned.
+    fn fetch_repository(&self, name: &RepositoryName) -> Result<Option<Repository>>;
+
+    /// Used by update_all_crates.
+    ///
+    /// The returned struct will contain all the information needed for `RepositoriesUpdater` to
+    /// update repositories that are still present and delete the missing ones.
+    fn fetch_repositories(&self, ids: &[String]) -> Result<FetchRepositoriesResult>;
+}
+
+pub struct Repository {
+    pub id: String,
+    pub name_with_owner: String,
+    pub description: Option<String>,
+    pub last_activity_at: Option<DateTime<Utc>>,
+    pub stars: i64,
+    pub forks: i64,
+    pub issues: i64,
+}
+
+#[derive(Default)]
+pub struct FetchRepositoriesResult {
+    pub present: HashMap<String, Repository>,
+    pub missing: Vec<String>,
+}
+
+pub struct RepositoryStatsUpdater {
+    updaters: Vec<Box<dyn RepositoryForge + Send + Sync>>,
+}
+
+impl fmt::Debug for RepositoryStatsUpdater {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "RepositoryStatsUpdater {{ {:?} }}",
+            self.updaters.iter().map(|u| u.host()).collect::<Vec<_>>()
+        )
+    }
+}
+
+impl RepositoryStatsUpdater {
+    pub fn new(config: &Config) -> Self {
+        let mut updaters: Vec<Box<dyn RepositoryForge + Send + Sync>> = Vec::with_capacity(3);
```

Comment:
I don't think `with_capacity` is needed there, and it's also easy to forget updating it when adding more forges.

Reply:
Switching to `new` then.